ssl/quic/quic_wire.c

```diff
@@ -0,0 +1,776 @@
+            || !WPACKET_quic_write_vlint(pkt, f->error_code))
+        return 0;
+
+    if (!f->is_app && !WPACKET_quic_write_vlint(pkt, f->frame_type))
+        return 0;
+
+    if (!WPACKET_quic_write_vlint(pkt, f->reason_len)
+            || !WPACKET_memcpy(pkt, f->reason, f->reason_len))
+        return 0;
+
+    return 1;
+}
+
+int ossl_quic_wire_encode_frame_handshake_done(WPACKET *pkt)
+{
+    return encode_frame_hdr(pkt, OSSL_QUIC_FRAME_TYPE_HANDSHAKE_DONE);
+}
+
+uint8_t *ossl_quic_wire_encode_transport_param_bytes(WPACKET *pkt,
+                                                     uint64_t id,
+                                                     const uint8_t *value,
+                                                     size_t value_len)
+{
+    uint8_t *b = NULL;
+
+    if (!WPACKET_quic_write_vlint(pkt, id)
+            || !WPACKET_quic_write_vlint(pkt, value_len)
+            || !WPACKET_allocate_bytes(pkt, value_len, (unsigned char **)&b))
+        return NULL;
+
+    if (value != NULL)
+        memcpy(b, value, value_len);
+
+    return b;
+}
+
+int ossl_quic_wire_encode_transport_param_int(WPACKET *pkt,
+                                              uint64_t id,
+                                              uint64_t value)
+{
+    if (!WPACKET_quic_write_vlint(pkt, id)
+            || !WPACKET_quic_write_vlint(pkt, ossl_quic_vlint_encode_len(value))
+            || !WPACKET_quic_write_vlint(pkt, value))
+        return 0;
+
+    return 1;
+}
+
+/*
+ * QUIC Wire Format Decoding
+ * =========================
+ */
+int ossl_quic_wire_peek_frame_header(PACKET *pkt, uint64_t *type)
+{
+    return PACKET_peek_quic_vlint(pkt, type);
+}
+
+int ossl_quic_wire_skip_frame_header(PACKET *pkt, uint64_t *type)
+{
+    return PACKET_get_quic_vlint(pkt, type);
+}
+
+static int expect_frame_header_mask(PACKET *pkt,
+                                    uint64_t expected_frame_type,
+                                    uint64_t mask_bits,
+                                    uint64_t *actual_frame_type)
+{
+    uint64_t actual_frame_type_;
+
+    if (!ossl_quic_wire_skip_frame_header(pkt, &actual_frame_type_)
+            || (actual_frame_type_ & ~mask_bits) != expected_frame_type)
+        return 0;
+
+    if (actual_frame_type != NULL)
+        *actual_frame_type = actual_frame_type_;
+
+    return 1;
+}
+
+static int expect_frame_header(PACKET *pkt, uint64_t expected_frame_type)
+{
+    uint64_t actual_frame_type;
+
+    if (!ossl_quic_wire_skip_frame_header(pkt, &actual_frame_type)
+            || actual_frame_type != expected_frame_type)
+        return 0;
+
+    return 1;
+}
+
+int ossl_quic_wire_decode_frame_ack(PACKET *pkt,
+                                    uint32_t ack_delay_exponent,
+                                    OSSL_ACKM_ACK *ack,
+                                    uint64_t *total_ranges) {
+    uint64_t frame_type, largest_ackd, ack_delay_raw,
+             ack_range_count, first_ack_range, start, end, i;
+
```

Comment:
Strictly yes. However we need to advance over the rest of the fields so the PACKET's cursor is at the start of the next frame. "If you pass NULL, the PACKET is in an undefined position on return" is not very helpful (even if in practice I expect users passing NULL will make a copy of the PACKET structure and call it again with the original PACKET structure).

Reply:
How do we envisage `ossl_quic_wire_decode_from_ack` and passing a NULL `ack` actually being used? I kind of envisaged calling it once with a NULL `ack`, then allocating the structure with the appropriate ranges and then calling it again with the same `pkt`...is that not how it works? If so we don't really need to advance the packet...although I suppose it would be a bit weird to advance in some cases but not others...

Reply:
Three other possibilities worthy of consideration:

1) Have the `ossl_quic_wire_decode_from_ack` function do the allocation itself

2) Have a separate function that just finds the `total_ranges` value without advancing the PACKET

3) Once `total_ranges` is known the length of the frame can be easily calculated. You could just do `PACKET_forward()` by the calculated amount.

Reply:
You can already do this. If the `num_ack_ranges` in the passed structure is inadequate, as many ranges as possible will be decoded. But you can detect that more were available by examining `*total_ranges`.

Still, a function to get only the number of ranges without advancing the PACKET or decoding the other fields should be slightly cheaper, so I'll add this.